core/include/detray/coordinates/cylindrical2.hpp

```diff
@@ -69,6 +84,90 @@ struct cylindrical2 : public coordinate_base<cylindrical2, transform3_t> {
         return trf.point_to_global(point3{x, y, z});
     }
 
+    template <typename mask_t>
+    DETRAY_HOST_DEVICE inline vector3 normal(const transform3_t &trf3,
+                                             const mask_t &mask,
+                                             const point3 &pos,
+                                             const vector3 &dir) const {
+        const point2 local2 = this->global_to_local(trf3, pos, dir);
+        const scalar_type r = mask[0];
+        const scalar_type phi = local2[0] / r;
+
+        // normal vector in local coordinate
+        const vector3 local_normal{std::cos(phi), std::sin(phi), 0};
+
+        // normal vector in global coordinate
+        return trf3.rotation() * local_normal;
+    }
+
+    template <typename mask_t>
+    DETRAY_HOST_DEVICE inline rotation_matrix reference_frame(
+        const transform3_t &trf3, const mask_t &mask, const point3 &pos,
+        const vector3 &dir) const {
+
+        rotation_matrix rot = matrix_operator().template zero<3, 3>();
+
+        // y axis of the new frame is the z axis of cylindrical coordinate
+        const auto new_yaxis =
+            matrix_operator().template block<3, 1>(trf3.matrix(), 0, 2);
+
+        // z axis of the new frame is the vector normal to the cylinder surface
+        const vector3 new_zaxis = normal(trf3, mask, pos, dir);
```

Comment:
It is ACTS convention. In cylindrical coordinate, the reference frame is defined with the intersection point on the cylinder surface (x axis in rphi direction and yaxis in cylinder axis). Resetting the reference frame in this way, we can reuse the jacobian transformation function written for cartesian coordinate.

If we just use transform matrix as the reference frame, we cannot use the same codes because intersection point would not be on the reference frame.

Reply:
Ok, thanks. Is it only the cylinder2 surface? In that case, would it make sense to put the extraction of the rotation matrix in ```coordinate_base``` and have the classes that implement a special reference frame, e.g. cylinder2, overwrite it? If something like that is possible, that would be easier to read I think

Reply:
Hmm I think I prefer the current state.